src/utils/dom-tranform-stream.js

```diff
@@ -0,0 +1,142 @@
+import {Services} from '../services';
+import {devAssert, user} from '../log';
+import {removeNoScriptElements} from './dom-writer';
+
+export class DomTransformStream {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    const headDefer = new Deferred();
+    /** @const @private {!Promise<!Element>} */
+    this.headPromise_ = headDefer.promise;
+
+    /** @const @private {!function()} */
+    this.headResolver_ = headDefer.resolve;
+
+    const transferDefer = new Deferred();
+    /** @const @private {!Promise} */
+    this.bodyTransferPromise_ = transferDefer.promise;
+
+    /** @const @private {!function()} */
+    this.bodyTransferResolver_ = transferDefer.resolve;
+
+    /** @private {?Element} */
+    this.detachedBody_ = null;
+
+    /** @private {?Promise} */
+    this.currentChunkTransferPromise_ = null;
+
+    /** @private {boolean} */
+    this.shouldTransfer_ = false;
+
+    /** @private {boolean} */
+    this.mergeScheduled_ = false;
+
+    /** @const @private */
+    this.vsync_ = Services.vsyncFor(win);
+  }
+
+  /**
+   * Callback passed into DetachedDomStream constructor. Receives updated
+   * document everytime a new chunk is written.
+   * Resolves headPromise when body is available, and streams to body iff
+   * bodyTransfer() has been called.
+   * @param {!Document} detachedDoc
+   */
+  onChunk(detachedDoc) {
+    // <body> is newly formed.
+    if (!this.detachedBody_ && detachedDoc.body) {
+      this.detachedBody_ = detachedDoc.body;
+      this.headResolver_(detachedDoc.head);
+    }
+
+    // If bodyTransfer has already been called, keep transferring on new chunks.
+    if (this.shouldTransfer_) {
+      this.transferBodyChunk_();
+    }
+  }
+
+  /**
+   * Callback passed into DetachedDomStream constructor. Called with complete
+   * doc when stream is closed.
+   * Schedules final transfer, then resovles body complete promise.
+   * @param {!Document} unusedCompleteDoc
+   */
+  onEnd(unusedCompleteDoc) {
+    this.transferBodyChunk_().then(this.bodyTransferResolver_);
+  }
+
+  /**
+   * Promise that will resolve with <head> when available.
+   * @return {!Promise<!Element>}
+   */
+  waitForHead() {
+    return this.headPromise_;
+  }
+
+  /**
+   * Start the body transfer process. Should only be called once.
+   * @param {!Element} target DOM element to be appended to.
+   * @return {!Promise} resolves when doc has been fully transferred.
+   */
+  transferBody(target) {
+    user().assertElement(
+      target,
+      'No target given to DomTransformStream.transferBody'
+    );
+
+    devAssert(
+      !this.shouldTransfer_,
+      'DomTransformStream.transferBody should only be called once'
+    );
+
+    this.shouldTransfer_ = true;
+    this.target_ = target;
+
+    this.transferBodyChunk_();
+
+    return this.bodyTransferPromise_;
+  }
+
+  /**
+   * Transfers available body elements in vsync cycle.
+   * @return {!Promise}
+   */
+  transferBodyChunk_() {
+    if (this.mergeScheduled_) {
```

Comment:
Can we combine these two vars `mergeScheduled_` and `currentChunkTransferPromise_` into a one - just the promise and use it's nullability for the test. Use of both is completely constrained inside this method.

Reply:
done.